src/LoginPage.cc

```diff
@@ -142,63 +177,111 @@ void LoginPage::loginError(QString error)
+		matrixid_spinner_->show();
+	}
 
-		QString home_server = custom_domain_.isEmpty()
-					      ? matrixid_input_->text().split(":").at(1)
-					      : custom_domain_;
+	server_layout_->invalidate();
+	matrixid_layout_->invalidate();
+}
 
-		client_->setServer(home_server);
-		client_->login(user, password);
-	}
+void LoginPage::onServerAddressEntered()
+{
+	guessing_server = false;
+	error_label_->setText("");
+	client_->setServer(server_input_->text());
+	client_->versions();
+
+	server_error_->hide();
+	server_success_->hide();
+	server_spinner_->show();
+	server_layout_->invalidate();
+	matrixid_layout_->invalidate();
 }
 
-void LoginPage::showSettingsModal()
+void LoginPage::versionError(QString error)
 {
-	if (login_settings_ == nullptr) {
-		login_settings_ = new LoginSettings(this);
-		connect(login_settings_, &LoginSettings::closing, this, &LoginPage::closeSettingsModal);
+	// Matrix homeservers are often kept on a subdomain called 'matrix'
+	// so let's try that next, unless the address was set explicitly
+	QUrl current_server = client_->getHomeServer();
+	if (guessing_server && !current_server.host().startsWith("matrix")) {
+		error_label_->setText("");
+		current_server.setHost(QString("matrix.")+current_server.host());
+		server_input_->setText(current_server.host());
+		client_->setServer(current_server.host());
+		client_->versions();
+		return;
 	}
 
-	if (settings_modal_ == nullptr) {
-		settings_modal_ = new OverlayModal(this, login_settings_);
-		settings_modal_->setDuration(100);
-		settings_modal_->setColor(QColor(55, 55, 55, 170));
+	error_label_->setText(error);
+	server_input_->show();
+
+	server_input_->setVisible(true);
+	server_error_->show();
+	server_spinner_->hide();
+	matrixid_spinner_->hide();
+
+	server_layout_->invalidate();
+	matrixid_layout_->invalidate();
+}
+
+void LoginPage::versionSuccess()
+{
+	if (server_input_->isVisible()) {
+		server_spinner_->hide();
+		matrixid_spinner_->hide();
+		server_success_->show();
+	} else {
+		matrixid_spinner_->hide();
+		matrixid_success_->show();
 	}
 
-	settings_modal_->fadeIn();
+	server_layout_->invalidate();
+	matrixid_layout_->invalidate();
 }
 
-void LoginPage::closeSettingsModal(const QString &server)
+void LoginPage::onLoginButtonClicked()
 {
-	custom_domain_ = server;
-	settings_modal_->fadeOut();
+	error_label_->setText("");
+
+	if (!matrixid_input_->hasAcceptableInput()) {
+		loginError("Invalid Matrix ID");
+	} else if (password_input_->text().isEmpty()) {
+		loginError("Empty password");
+	} else {
+		QString user = matrixid_input_->text().split(":").at(0).split("@").at(1);
+		QString password = password_input_->text();
+		client_->login(user, password);
+	}
 }
 
 void LoginPage::reset()
 {
 	matrixid_input_->clear();
 	password_input_->clear();
+	server_input_->clear();
 
-	if (settings_modal_ != nullptr) {
-		settings_modal_->deleteLater();
-		settings_modal_ = nullptr;
-	}
+	server_success_->hide();
+	server_error_->hide();
+	server_spinner_->hide();
```

Comment:
`reset()` is called when we're done with the login page and there is no reason keeping optional elements visible. For example if the user calls logout and tries to login again he will see the server input.